### PR TITLE
OCPBUGS-35453: [release-4.13] internal service changed fix target port logic

### PR DESCRIPTION
**Pull request by candita**
This is a manual cherry-pick of #1072.

These two commits were not backported due to compatibility issues:
0cbd86a Test_desiredLoadBalancer: Add assertions
87975f7 Test_deploymentConfigChanged: Use k8s.io/utils/ptr

git cherry-pick 71eead5 5d0939a  14daecd f2060e0 826745c 50e8a73 eb1623e caeca65 2aed24f

fix merge conflict in test/e2e/all_test.go
(go mod tidy; go mod vendor; make update; make verify; make build; make test)
git cherry-pick --continue

fix merge conflict in pkg/operator/controller/ingress/load_balancer_service_test.go
(go mod tidy; go mod vendor; make update; make verify; make build; make test)

git cherry-pick --continue
(go mod tidy; go mod vendor; make update; make verify; make build; make test)
git cherry-pick --continue
Fix merge conflict in pkg/operator/controller/ingress/load_balancer_service_test.go

git cherry-pick continue
Fix merge conflict in test/e2e/operator_test.go
(go mod tidy; go mod vendor; make update; make verify; make build; make test)

Test failed, add module changes for backport and fix test:
go.mod
go.sum
pkg/manifests/bindata.go
pkg/operator/controller/ingress/deployment_test.go
vendor/k8s.io/utils/integer/integer.go
vendor/k8s.io/utils/pointer/pointer.go
vendor/k8s.io/utils/ptr/OWNERS
vendor/k8s.io/utils/ptr/README.md
vendor/k8s.io/utils/ptr/ptr.go
vendor/k8s.io/utils/trace/trace.go
vendor/modules.txt
git commit --amend

git cherry-pick 1f9795b
(go mod tidy; go mod vendor; make verify; make update; make build; make test)
git add /pkg/manifests/bindata.go
git commit --amend

git cherry-pick 77595d0
(go mod tidy; go mod vendor; make verify; make update; make build; make test)
git add /pkg/manifests/bindata.go
git commit --amend

### CODE REVIEW

**Comment by ShudiLi** [comment]
/label qe-approved
tested it with 4.13.0-0.ci.test-2024-06-18-014916-ci-ln-5b4jc4k-latest
```
1.
% oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-0.ci.test-2024-06-18-014916-ci-ln-5b4jc4k-latest   True        False         89s     Cluster version is 4.13.0-0.ci.test-2024-06-18-014916-ci-ln-5b4jc4k-latest

2.
% oc create -f ingress-hostnetwork.yaml
ingresscontroller.operator.openshift.io/int1 created
% oc -n openshift-ingress get pods
NAME                             READY   STATUS    RESTARTS   AGE
router-default-6954d5d76-njf74   1/1     Running   0          29m
router-default-6954d5d76-rpjn5   1/1     Running   0          29m
router-int1-d79944968-sgbbp      1/1     Running   0          20s

3.
% oc -n openshift-ingress get svc router-internal-int1 -o=jsonpath="{.spec.ports}"
[{"name":"http","port":80,"protocol":"TCP","targetPort":"http"},{"name":"https","port":443,"protocol":"TCP","targetPort":"https"},{"name":"metrics","port":1936,"protocol":"TCP","targetPort":"metrics"}]%  
                                                                                                                                                        
4.
% oc -n openshift-ingress get ep router-internal-int1 -o=jsonpath="{.subsets[0].ports}"
[{"name":"metrics","port":23936,"protocol":"TCP"},{"name":"https","port":23443,"protocol":"TCP"},{"name":"http","port":23080,"protocol":"TCP"}]%          

5.        
% oc -n openshift-ingress get deployment router-int1 -oyaml | grep -E -A9 "ivenessProbe:|readinessProbe:|startupProbe:"
        livenessProbe:
          failureThreshold: 3
          httpGet:
            host: localhost
            path: /healthz
            port: 23936
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          terminationGracePeriodSeconds: 10
--
        readinessProbe:
          failureThreshold: 3
          httpGet:
            host: localhost
            path: /healthz/ready
            port: 23936
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
--
        startupProbe:
          failureThreshold: 120
          httpGet:
            host: localhost
            path: /healthz/ready
            port: 23936
            scheme: HTTP
          periodSeconds: 1
          successThreshold: 1
          timeoutSeconds: 1

6.
% oc -n openshift-ingress get deployment router-int1 -oyaml | grep -A6 " - configMap:"
      - configMap:
          defaultMode: 420
          items:
          - key: service-ca.crt
            path: service-ca.crt
          name: service-ca-bundle
          optional: false

7.
% oc -n openshift-ingress get svc router-internal-int1 -o=jsonpath="{.spec.sessionAffinity}" 
None%                                                                                                                                                                             

8.
% oc create -f ingress-internal.yaml 
ingresscontroller.operator.openshift.io/int2 created

9.
% print $(oc -n openshift-ingress get svc router-int2 -o=jsonpath="{.spec.type}\n{.spec.sessionAffinity}")
LoadBalancer
None

10.
shudi@Shudis-MacBook-Pro NE982 % oc -n openshift-ingress-operator edit ingresscontroller int1
ingresscontroller.operator.openshift.io/int1 edited

11.
shudi@Shudis-MacBook-Pro NE982 %  oc -n openshift-ingress get ep router-internal-int1 -o=jsonpath="{.subsets[0].ports}"
[{"name":"https","port":24443,"protocol":"TCP"},{"name":"http","port":24080,"protocol":"TCP"},{"name":"metrics","port":24936,"protocol":"TCP"}]%                           

12.
shudi@Shudis-MacBook-Pro NE982 % oc -n openshift-ingress get deployment router-int1 -oyaml | grep -E -A9 "ivenessProbe:|readinessProbe:|startupProbe:"
        livenessProbe:
          failureThreshold: 3
          httpGet:
            host: localhost
            path: /healthz
            port: 24936
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          terminationGracePeriodSeconds: 10
--
        readinessProbe:
          failureThreshold: 3
          httpGet:
            host: localhost
            path: /healthz/ready
            port: 24936
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
--
        startupProbe:
          failureThreshold: 120
          httpGet:
            host: localhost
            path: /healthz/ready
            port: 24936
            scheme: HTTP
          periodSeconds: 1
          successThreshold: 1
          timeoutSeconds: 1

13.
% rm ingress-operator 
shudi@Shudis-MacBook-Pro key100 % oc -n openshift-ingress-operator -c ingress-operator rsync ingress-operator-c8bc449bd-qp85t:/usr/bin/ingress-operator .
receiving file list ... done
ingress-operator

14.
% go version -m ingress-operator | grep k8s.io/utils
	dep	k8s.io/utils	v0.0.0-20240502163921-fe8a2dddb1d0
```

**Comment by Miciah** [comment]
@candita, it looks like you're backporting #884 too—is that your intention?

**Comment by Miciah** [comment]
Did you mean to drop the `assert.Equal(t, "None", string(svc.Spec.SessionAffinity))` in `Test_desiredLoadBalancerService` from https://github.com/openshift/cluster-ingress-operator/commit/f2060e0c0440d9ac79e0c456b564ef0047709007?

**Comment by Miciah** [comment]
> This is a manual backport of #1086.
> 
> These two commits were not backported due to compatibility issues: [0cbd86a](https://github.com/openshift/cluster-ingress-operator/commit/0cbd86a4438c45be8c2917b781503b06f52bd55e) Test_desiredLoadBalancer: Add assertions [87975f7](https://github.com/openshift/cluster-ingress-operator/commit/87975f7833badbe8cbb9942eaf0a46c3c9481f9a) Test_deploymentConfigChanged: Use k8s.io/utils/ptr
> 
> git cherry-pick [71eead5](https://github.com/openshift/cluster-ingress-operator/commit/71eead5286597657de281306876b5d94dea7e4ad) [5d0939a](https://github.com/openshift/cluster-ingress-operator/commit/5d0939ad56e239919acfd962882ab7c076b4b7d9) [14daecd](https://github.com/openshift/cluster-ingress-operator/commit/14daecdb9ca49a7742afd922a8a9b55acf6935a6) [f2060e0](https://github.com/openshift/cluster-ingress-operator/commit/f2060e0c0440d9ac79e0c456b564ef0047709007) [826745c](https://github.com/openshift/cluster-ingress-operator/commit/826745c3a1db822c0538f3322e98b33bd628ed13) [50e8a73](https://github.com/openshift/cluster-ingress-operator/commit/50e8a730070d8f0864694df476da4c9fdf2a168b) [eb1623e](https://github.com/openshift/cluster-ingress-operator/commit/eb1623e71ca636ab3a3ecf2b39ab59e42209f5b8) [caeca65](https://github.com/openshift/cluster-ingress-operator/commit/caeca655be660dd93be63d3bc370c8616deb8641) [2aed24f](https://github.com/openshift/cluster-ingress-operator/commit/2aed24f184b7f98124547d293a5a316e3eeca535)

Did you mean to cherry-pick the commits from the release-4.16 branch, or did you mean to cherry-pick commits from the release-4.14 branch?

**Comment by Miciah** [comment]
The 4.13 backport looks the same as the 4.14 backport modulo changes to test code and generated files.  I do have some questions about the changes to the tests and possible loss of test coverage.  However, the functional changes look fine.  

/label backport-risk-assessed

**Comment by Miciah** [comment]
> > Did you mean to drop the `assert.Equal(t, "None", string(svc.Spec.SessionAffinity))` in `Test_desiredLoadBalancerService` from [f2060e0](https://github.com/openshift/cluster-ingress-operator/commit/f2060e0c0440d9ac79e0c456b564ef0047709007)?
> 
> Yes, it was giving a null pointer error in tests because that field wasn't populated.

Which field?  Neither `svc.Spec` nor `svc.Spec.SessionAffinity` is a pointer.  Anyway, if the field isn't populated, then doesn't that mean that the test found something wrong?

**Comment by candita** [comment]
> > > Did you mean to drop the `assert.Equal(t, "None", string(svc.Spec.SessionAffinity))` in `Test_desiredLoadBalancerService` from [f2060e0](https://github.com/openshift/cluster-ingress-operator/commit/f2060e0c0440d9ac79e0c456b564ef0047709007)?
> > 
> > 
> > Yes, it was giving a null pointer error in tests because that field wasn't populated.
> 
> Which field? Neither `svc.Spec` nor `svc.Spec.SessionAffinity` is a pointer. Anyway, if the field isn't populated, then doesn't that mean that the test found something wrong?

It fails on the test "nodePort service for aws platform" because `svc` is nil, and that is nil because we don't return a desired service when endpoint publishing strategy is `LoadBalancerServiceStrategyType`.

**Comment by candita** [comment]
openshift cluster install failed overall
/test e2e-aws-ovn

**Comment by candita** [comment]
Cluster install failure:

> level=info msg=Cluster operator insights SCAAvailable is False with Forbidden: Failed to pull SCA certs from https://api.openshift.com/api/accounts_mgmt/v1/certificates: OCM API https://api.openshift.com/api/accounts_mgmt/v1/certificates returned HTTP 403: {"code":"ACCT-MGMT-11","href":"/api/accounts_mgmt/v1/errors/11","id":"11","kind":"Error","operation_id":"8a5f8a5f-c132-496d-b49c-021908e2ef40","reason":"Account with ID 2DUeKzzTD9ngfsQ6YgkzdJn1jA4 denied access to perform create on Certificate with HTTP call POST /api/accounts_mgmt/v1/certificates"}

/test e2e-aws-ovn-upgrade

**Comment by candita** [comment]
> > > > Did you mean to drop the `assert.Equal(t, "None", string(svc.Spec.SessionAffinity))` in `Test_desiredLoadBalancerService` from [f2060e0](https://github.com/openshift/cluster-ingress-operator/commit/f2060e0c0440d9ac79e0c456b564ef0047709007)?
> > > 
> > > 
> > > Yes, it was giving a null pointer error in tests because that field wasn't populated.
> > 
> > 
> > Which field? Neither `svc.Spec` nor `svc.Spec.SessionAffinity` is a pointer. Anyway, if the field isn't populated, then doesn't that mean that the test found something wrong?
> 
> It fails on the test "nodePort service for aws platform" because `svc` is nil, and that is nil because we don't return a desired service when endpoint publishing strategy is `LoadBalancerServiceStrategyType`.

I've updated the PR with a solution I mentioned in slack. PTAL @Miciah when you can.

**Comment by candita** [comment]
> @candita, it looks like you're backporting #884 too—is that your intention?

As discussed in slack, #884 was not intentionally included, but included as a byproduct of another commit.  If there is something you'd like me to remove, please let me know.

**Comment by candita** [comment]
> > This is a manual backport of #1086.
> > These two commits were not backported due to compatibility issues: [0cbd86a](https://github.com/openshift/cluster-ingress-operator/commit/0cbd86a4438c45be8c2917b781503b06f52bd55e) Test_desiredLoadBalancer: Add assertions [87975f7](https://github.com/openshift/cluster-ingress-operator/commit/87975f7833badbe8cbb9942eaf0a46c3c9481f9a) Test_deploymentConfigChanged: Use k8s.io/utils/ptr
> > git cherry-pick [71eead5](https://github.com/openshift/cluster-ingress-operator/commit/71eead5286597657de281306876b5d94dea7e4ad) [5d0939a](https://github.com/openshift/cluster-ingress-operator/commit/5d0939ad56e239919acfd962882ab7c076b4b7d9) [14daecd](https://github.com/openshift/cluster-ingress-operator/commit/14daecdb9ca49a7742afd922a8a9b55acf6935a6) [f2060e0](https://github.com/openshift/cluster-ingress-operator/commit/f2060e0c0440d9ac79e0c456b564ef0047709007) [826745c](https://github.com/openshift/cluster-ingress-operator/commit/826745c3a1db822c0538f3322e98b33bd628ed13) [50e8a73](https://github.com/openshift/cluster-ingress-operator/commit/50e8a730070d8f0864694df476da4c9fdf2a168b) [eb1623e](https://github.com/openshift/cluster-ingress-operator/commit/eb1623e71ca636ab3a3ecf2b39ab59e42209f5b8) [caeca65](https://github.com/openshift/cluster-ingress-operator/commit/caeca655be660dd93be63d3bc370c8616deb8641) [2aed24f](https://github.com/openshift/cluster-ingress-operator/commit/2aed24f184b7f98124547d293a5a316e3eeca535)
> 
> Did you mean to cherry-pick the commits from the release-4.16 branch, or did you mean to cherry-pick commits from the release-4.14 branch?

I updated the first line of the description to:

> This is a manual cherry-pick of https://github.com/openshift/cluster-ingress-operator/pull/1072.

**Comment by candita** [comment]
maxAllowed disruption of 0s is not alway met, hopefully this is transient:
> ci-cluster-network-liveness-reused-connections was unreachable during disruption testing for at least 1s of 1h31m22s (maxAllowed=0s)

/test e2e-aws-ovn-serial

**Comment by Miciah** [comment]
> > @candita, it looks like you're backporting #884 too—is that your intention?
> 
> As discussed in slack, #884 was not intentionally included, but some changes from it were included as a byproduct of another commit. If there is something you'd like me to remove, please let me know.

It is a bit confusing, it does make reviewing a bit more difficult, and it may make future backports more confusing as well, but as this PR is urgent, we can go with it.

File: pkg/operator/controller/canary/route_test.go [hunk 1]
```diff
@@ -134,10 +134,13 @@ func TestCanaryRouteChanged(t *testing.T) {
...
 		if changed, updated := canaryRouteChanged(original, mutated); changed != tc.expect {
-			t.Errorf("%s, expect canaryRouteChanged to be %t, got %t", tc.description, tc.expect, changed)
+			t.Errorf("expected canaryRouteChanged to be %t, got %t", tc.expect, changed)
 		} else if changed {
+			if updatedChanged, _ := canaryRouteChanged(original, updated); !updatedChanged {
+				t.Error("canaryRouteChanged reported changes but did not make any update")
+			}
 			if changedAgain, _ := canaryRouteChanged(mutated, updated); changedAgain {
-				t.Errorf("%s, canaryRouteChanged does not behave as a fixed point function", tc.description)
+				t.Error("canaryRouteChanged does not behave as a fixed point function")
```
Comment (Miciah) [inline pkg/operator/controller/canary/route_test.go:143]: Here, `t.Run` _isn't_ added, but the formats are changed to omit `tc.description`.  Very confusing.

File: pkg/operator/controller/ingress/deployment_test.go [hunk 2]
```diff
@@ -1740,13 +1740,17 @@ func TestDeploymentConfigChanged(t *testing.T) {
...
 		if changed, updated := deploymentConfigChanged(&original, mutated); changed != tc.expect {
-			t.Errorf("%s, expect deploymentConfigChanged to be %t, got %t", tc.description, tc.expect, changed)
+			t.Errorf("expect deploymentConfigChanged to be %t, got %t", tc.expect, changed)
 		} else if changed {
+			if updatedChanged, _ := deploymentConfigChanged(&original, updated); !updatedChanged {
+				t.Error("deploymentConfigChanged reported changes but did not make any update")
+			}
 			if changedAgain, _ := deploymentConfigChanged(mutated, updated); changedAgain {
-				t.Errorf("%s, deploymentConfigChanged does not behave as a fixed point function", tc.description)
+				t.Error("deploymentConfigChanged does not behave as a fixed point function")
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment_test.go:1875]: Here too, `t.Run` _isn't_ added, but the strings are updated as if it were being added.

File: pkg/operator/controller/ingress/load_balancer_service_test.go [hunk 3]
```diff
@@ -466,6 +466,10 @@ func TestDesiredLoadBalancerService(t *testing.T) {
 				t.Errorf("local-with-fallback annotation check for test %q failed: %v", tc.description, err)
 			}
 		}
+		// Only make this assertion if we expect a service.
+		if tc.expect {
+			assert.Equal(t, "None", string(svc.Spec.SessionAffinity))
```
Comment (Miciah) [inline pkg/operator/controller/ingress/load_balancer_service_test.go:471]: Historical note: https://github.com/openshift/cluster-ingress-operator/pull/886/commits/a3bb33ec0e7fcae199024d38dbb113396130a4ef adds an `if !tc.expect || !haveSvc { return }` guard.  The `if tc.expect` needed to be added here as #886 wasn't backported to the release-4.13 branch.

File: pkg/operator/controller/ingress/nodeport_service_test.go [hunk 4]
```diff
@@ -329,10 +329,13 @@ func TestNodePortServiceChanged(t *testing.T) {
...
 		if changed, updated := nodePortServiceChanged(&original, mutated); changed != tc.expect {
-			t.Errorf("%s, expect nodePortServiceChanged to be %t, got %t", tc.description, tc.expect, changed)
+			t.Errorf("expect nodePortServiceChanged to be %t, got %t", tc.expect, changed)
 		} else if changed {
+			if updatedChanged, _ := nodePortServiceChanged(&original, updated); !updatedChanged {
+				t.Error("nodePortServiceChanged reported changes but did not make any update")
+			}
 			if changedAgain, _ := nodePortServiceChanged(mutated, updated); changedAgain {
-				t.Errorf("%s, nodePortServiceChanged does not behave as a fixed point function", tc.description)
+				t.Error("nodePortServiceChanged does not behave as a fixed point function")
```
Comment (Miciah) [inline pkg/operator/controller/ingress/nodeport_service_test.go:340]: Here too, no `t.Run`, yet strings are updated.

File: pkg/operator/controller/ingressclass/ingressclass_test.go [hunk 5]
```diff
@@ -177,10 +177,13 @@ func TestIngressClassChanged(t *testing.T) {
...
 		if changed, updated := ingressClassChanged(&original, mutated); changed != tc.expect {
-			t.Errorf("%s, expect ingressClassChanged to be %t, got %t", tc.description, tc.expect, changed)
+			t.Errorf("expect ingressClassChanged to be %t, got %t", tc.expect, changed)
 		} else if changed {
+			if updatedChanged, _ := ingressClassChanged(&original, updated); !updatedChanged {
+				t.Error("ingressClassChanged reported changes but did not make any update")
+			}
 			if changedAgain, _ := ingressClassChanged(mutated, updated); changedAgain {
-				t.Errorf("%s, ingressClassChanged does not behave as a fixed point function", tc.description)
+				t.Error("ingressClassChanged does not behave as a fixed point function")
```
Comment (Miciah) [inline pkg/operator/controller/ingressclass/ingressclass_test.go:186]: And here too, no `t.Run`, yet strings are updated.  So actually in most cases, `t.Run` wasn't introduced, but the strings were updated.  

Clearly, something weird happened with the conflict resolution, but I am fairly confident that tests will still fail when they ought to (though some tests will give unhelpful output when they fail).

**Comment by Miciah** [comment]
The automatic conflict did some really weird things in the test code, but the functional changes are fine, and the tests look like they will still when they ought to.  
/approve
/lgtm

**Comment by candita** [comment]
This is a new type of cluster installation failure:

> 0001-01-01T00:00:00Z 0x : 0/36 nodes are available: 1 Insufficient cpu, 1 node(s) had untolerated taint {node-role.kubernetes.io/ci-longtests-worker: ci-longtests-worker}, 16 node(s) had untolerated taint {node-role.kubernetes.io/ci-tests-worker: ci-tests-worker}, 2 node(s) had untolerated taint {node-role.kubernetes.io/infra: }, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }, 4 node(s) didn't match Pod's node affinity/selector, 9 node(s) had untolerated taint {node-role.kubernetes.io/ci-prowjobs-worker: ci-prowjobs-worker}. preemption: 0/36 nodes are available: 1 No preemption victims found for incoming pod, 35 Preemption is not helpful for scheduling..

**Comment by candita** [comment]
I suspect there is something wrong with this test, based on the error reports.

/override e2e-gcp-operator

**Pull request merged**